src/net.cpp

```diff
@@ -2837,16 +2837,109 @@ void CConnman::PushMessage(CNode* pnode, CSerializedNetMsg&& msg)
 bool CConnman::ForNode(NodeId id, std::function<bool(CNode* pnode)> func)
 {
     CNode* found = nullptr;
-    LOCK(cs_vNodes);
-    for (auto&& pnode : vNodes) {
-        if(pnode->GetId() == id) {
-            found = pnode;
-            break;
+    {
+        LOCK(cs_vNodes);
+        for (auto&& pnode : vNodes) {
+            if(pnode->GetId() == id) {
+                found = pnode;
+                found->AddRef();
+                break;
+            }
         }
     }
-    return found != nullptr && NodeFullyConnected(found) && func(found);
+    bool res = false;
+    if (found != nullptr) {
+        res = NodeFullyConnected(found) && func(found);
+        found->Release();
+    }
+    return res;
 }
 
+void CConnman::ForEachNode(std::function<void (CNode* pnode)> func)
+{
+    std::vector<CNode*> vNodes_copy;
```

Comment:
These should just call ForEachNodeThen with a null post functions so this code doesn't have to be repeated so many times. Again see eeb85c63cf880bc7344413b266c3af793d5cb8d2.